lib/puppet/ssl/verifier.rb

```diff
@@ -0,0 +1,132 @@
+require 'puppet/ssl'
+
+# Verify an SSL connection.
+#
+# @api private
+class Puppet::SSL::Verifier
+
+  FIVE_MINUTES_AS_SECONDS = 5 * 60
+
+  # Create a verifier using an `ssl_context`.
+  #
+  # @param [Puppet::SSL::SSLContext] ssl_context containing CA certs,
+  #   CRLs, etc needed to verify the server's certificate chain
+  def initialize(ssl_context)
+    @ssl_context = ssl_context
+    @errors = []
+  end
+
+  # Configure the `http` connection based on the current `ssl_context`.
+  #
+  # @param http [Net::HTTP] connection
+  # @api private
+  def setup_connection(http)
+    http.cert_store = @ssl_context[:store]
+    http.cert = @ssl_context[:client_cert]
+    http.key = @ssl_context[:private_key]
+    # default to VERIFY_PEER
+    http.verify_mode = if !@ssl_context[:verify_peer]
+                         OpenSSL::SSL::VERIFY_NONE
+                       else
+                         OpenSSL::SSL::VERIFY_PEER
+                       end
+    http.verify_callback = self
+  end
+
+  # This method is called if `Net::HTTP#start` raises an exception, which
+  # could be a result of an openssl error during cert verification, due
+  # to ruby's `Socket#post_connection_check`, or general SSL connection
+  # error.
+  #
+  # @param http [Net::HTTP] connection
+  # @param error [OpenSSL::SSL::SSLError] connection error
+  # @raise [Puppet::SSL::CertVerifyError] SSL connection failed due to a
+  #   verification error with the server's certificate or chain
+  # @raise [Puppet::Error] server hostname does not match certificate
+  # @raise [OpenSSL::SSL::SSLError] low-level SSL connection failure
+  # @api private
+  def handle_connection_error(http, error)
+    peer_cert = http.peer_cert
+    host = http.address
+
+    if peer_cert && !OpenSSL::SSL.verify_certificate_identity(peer_cert, host)
+      # if `SSLSocket#post_connection_check` raises (ruby < 2.4), then peer_cert will be valid
+      raise cert_mismatch_error(peer_cert, host)
+    elsif !@errors.empty?
+      err = @errors.first
```

Comment:
When the agent verifies the server's cert chain, ruby/openssl will call our `call` method below, at least once per cert. When running ruby >= 2.4, we need to check if one of the verification errors was due to a cert/hostname mismatch. I could check each verification error to see if it's a cert mismatch, but that seems not super useful. So I just check the first error, and either raise the `CertMismatchError` or raise the first error itself (which will be a `CertVerifyError`).

Reply:
So if the first error isn't a cert mismatch, we don't care if any other errors are a cert mismatch?

Reply:
The case where `@errors.length > 1` and there's a cert mismatch, is if there's something wrong with one of the CA certs in the chain and the server's cert doesn't match. Like if the agent tried to send a report to a third-party SSL server, and it doesn't trust the CA cert:

```puppet
$ bx puppet agent -t --report_server https://<wronghostname>`
```

In that case, we'll report on the first verification error will be the untrusted CA cert (since the chain is processed from root to leaf), and we'll raise that. It kind of doesn't matter that there was a later cert error, until the first issue is resolved.

Reply:
ohhhh, actually the `Verifier` returns false as soon as it gets an error, so there will only be one error